### PR TITLE
spirv-fuzz: Fuzzer pass to add composite types

### PR DESCRIPTION
**Pull request by afd**
Adds a fuzzer pass that randomly adds vector and matrix types not
already present in the module, and randomly adds structs with random
field types and arrays with random base types and sizes.  Other passes
will be able to create variables and ids using these types.

### CODE REVIEW

File: source/fuzz/fuzzer_pass_add_composite_types.cpp [hunk 1]
```diff
@@ -0,0 +1,137 @@
+// Copyright (c) 2019 Google LLC
```
Comment (paulthomson) [inline source/fuzz/fuzzer_pass_add_composite_types.cpp]: Date?

File: source/fuzz/fuzzer_pass_add_composite_types.h [hunk 2]
```diff
@@ -0,0 +1,61 @@
+// Copyright (c) 2019 Google LLC
```
Comment (paulthomson) [inline source/fuzz/fuzzer_pass_add_composite_types.h]: Date?

File: source/fuzz/fuzzer_pass_add_composite_types.cpp [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+    protobufs::TransformationSequence* transformations)
+    : FuzzerPass(ir_context, fact_manager, fuzzer_context, transformations) {}
+
+FuzzerPassAddCompositeTypes::~FuzzerPassAddCompositeTypes() = default;
+
+void FuzzerPassAddCompositeTypes::Apply() {
+  MaybeAddMissingVectorTypes();
+  MaybeAddMissingMatrixTypes();
+
+  // (3) Randomly interleave between adding struct and array composite types
```
Comment (paulthomson) [inline source/fuzz/fuzzer_pass_add_composite_types.cpp]: (3) ?

File: source/fuzz/fuzzer_pass_add_composite_types.cpp [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+      GetFuzzerContext()->GetChanceOfAddingAnotherStructField()));
+  ApplyTransformation(TransformationAddTypeStruct(
+      GetFuzzerContext()->GetFreshId(), field_type_ids));
+}
+
+uint32_t FuzzerPassAddCompositeTypes::ChooseScalarOrCompositeType() {
+  // Gather up all the possibly-relevant types.
+  std::vector<uint32_t> candidates;
+  for (auto& inst : GetIRContext()->types_values()) {
+    switch (inst.opcode()) {
```
Comment (paulthomson) [inline source/fuzz/fuzzer_pass_add_composite_types.cpp]: Could you have struct types here?

File: source/fuzz/fuzzer_pass_add_composite_types.h (hunk 2 above)
Comment (afd) [reply source/fuzz/fuzzer_pass_add_composite_types.h]: Fixed, thanks.

File: source/fuzz/fuzzer_pass_add_composite_types.cpp (hunk 1 above)
Comment (afd) [reply source/fuzz/fuzzer_pass_add_composite_types.cpp]: Fixed, thanks.

File: source/fuzz/fuzzer_pass_add_composite_types.cpp (hunk 3 above)
Comment (afd) [reply source/fuzz/fuzzer_pass_add_composite_types.cpp]: Oops (a remnant from before I refactored) - fixed.

File: source/fuzz/fuzzer_pass_add_composite_types.cpp (hunk 4 above)
Comment (afd) [reply source/fuzz/fuzzer_pass_add_composite_types.cpp]: Amazing catch, thanks - it was absolutely my intention to have structs.

**Pull request merged**